### PR TITLE
feat: implementing token-highlighter worker, analyzer, demo

### PR DESCRIPTION
**Pull request by asanth7**
## Summary
- Implements token highlighter worker/analyzer flow
- Adds demo wiring and reporting updates
- Documents algorithm + worker/analyzer contract
- View `TokenHighlighter.md` for more detailed information on structure, architecture, and contribution.

## Type of contribution
- [x] New worker
- [x] New analyzer
- [ ] Bug fix
- [ ] Other (describe below)

## Files modified
- `vllm_hook_plugins/vllm_hook_plugins/workers/highlighter_worker.py` (new worker)
- `vllm_hook_plugins/vllm_hook_plugins/analyzers/highlighter_analyzer.py` (new analyzer)
- `examples/demo_token_highlighter.py` (demo)
- `TokenHighlighter.md` (detailed description of contribution)
- `vllm_hook_plugins/vllm_hook_plugins/__init__.py` (registered worker/analyzer)
- `model_configs/token_highlighter/Qwen2-1.5B-Instruct.json` (model configs)

- [x] I have NOT modified `hook_llm.py`

## Plugin architecture checklist
- [x] New workers/analyzers are registered via `PluginRegistry` in `__init__.py`
- [x] New workers extend `V1Worker` (not `HookLLM`)
- [x] `hooks_on=(prefill, generate)` flag is set correctly for any new worker registration
- [x] Examples or notebooks are included for new features

## Testing
Tested with `examples/demo_token_highlighter.py` using a local Qwen2-1.5B-Instruct snapshot on a single GPU (tensor_parallel_size=1, max_model_len=512, NVIDIA RTX 5070, 8GB VRAM). For each prompt, the demo runs one hooked `llm.generate()` pass and one `llm.analyze()` pass, then prints applied driver tokens, optional analyzer-side driver tokens (if analysis spec differs), and top tokens by score. I validated that scores are non-zero under the autograd scorer path, soft-removal is applied to selected prompt tokens, and the analyzer can be rerun with different specs without recomputing gradients.

## Related issue
N/A

## Contribution acknowledgement
If this contribution is included in a future version of the vLLM-Hook technical report, would you like to be credited as a co-author?

- [x] Yes, please include me as a contributor
- [ ] No, thanks

If yes, please provide:
- **Name**: Arav Santhanam
- **Affiliation**: Carnegie Mellon University
- **One-sentence description of your contribution**: Adds a gradient-based Token Highlighter that identifies and softly suppresses prompt tokens driving affirmative jailbreak behavior, with worker-side mitigation and analyzer-side post-run inspection (proposed in ICX360).

### CODE REVIEW

**Comment by IRENEKO** [comment]
@asanth7 Thank you for the PR! Can you please prepare a notebook example of token-highlighter as well? 

A good reference is here: https://github.com/IBM/vLLM-Hook/blob/main/notebooks/demo_attntracker.ipynb (similar to your `examples/demo_token_highlighter.py` but run in a notebook.

File: .vscode/settings.json
Comment (IRENEKO) [inline .vscode/settings.json:1]: Can we remove this from the PR? You can keep the files locally for your own use.

File: examples/compare_token_highlighter_scorers.py
Comment (IRENEKO) [inline examples/compare_token_highlighter_scorers.py:1]: Is this script intended for users to try out as a use case example? If this is more like a sanity check code for developers, then let's consider removing it from the PR and keeping it locally for your own reference.

File: examples/demo_token_highlighter.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from transformers import AutoTokenizer
+
+mp.set_start_method("spawn", force=True)
+os.environ["VLLM_USE_V1"] = "1"
+os.environ["VLLM_WORKER_MULTIPROC_METHOD"] = "spawn"
+
+from vllm_hook_plugins import HookLLM
+
+
+class HighlighterDemoLLM(HookLLM):
```
Comment (IRENEKO) [inline examples/demo_token_highlighter.py]: Is this subclass needed?

File: examples/demo_token_highlighter.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    """Demo-only wrapper: maps highlighter config into worker env vars."""
+
+    def load_config(self, config_file: str):
+        super().load_config(config_file)
+        with open(config_file, "r") as f:
+            cfg = json.load(f)
+        highlighter = cfg.get("highlighter", {})
+        os.environ["VLLM_HIGHLIGHTER_TARGET_PHRASE"] = highlighter.get(
+            "target_phrase",
+            "Sure! I can help with that.",
```
Comment (IRENEKO) [inline examples/demo_token_highlighter.py]: Is this a duplicate of the config? I see you have defined target phrase in the config

File: examples/demo_token_highlighter.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        )
+        os.environ["VLLM_HIGHLIGHTER_SCORER"] = highlighter.get(
```
Comment (IRENEKO) [inline examples/demo_token_highlighter.py]: Can we put most (or even better, all) these configs into the config files? and add a description for each of the variable and their intended use. In the demo, we just load all configs from the config files and send them as environment variables if needed.

File: examples/demo_token_highlighter.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        model,
+        trust_remote_code=True,
+        local_files_only=True,
+    ).encode(target_phrase, add_special_tokens=False)
+    os.environ["VLLM_HIGHLIGHTER_TARGET_TOKEN_IDS"] = ",".join(
+        str(tok) for tok in target_ids)
+
+    llm = HighlighterDemoLLM(
+        model=model,
+        worker_name="tok_grads_soft_hook",
```
Comment (IRENEKO) [inline examples/demo_token_highlighter.py]: Where is tok_grads_soft_hook defined? somehow I don't see this worker in the __init__

File: examples/demo_token_highlighter.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        analyzer_name="token_highlighter",
+        config_file=cfg_path,
+        download_dir=cache_dir,
+        gpu_memory_utilization=0.75,
+        max_model_len=512,
+        trust_remote_code=True,
+        dtype=torch.float16,
+        enforce_eager=True,
+        enable_prefix_caching=False,
```
Comment (IRENEKO) [inline examples/demo_token_highlighter.py:45]: Is this a hard False or prefix caching is possible too here?

File: notebooks/demo_token_highlighter/demo_token_highlighter.ipynb
Comment (IRENEKO) [inline notebooks/demo_token_highlighter/demo_token_highlighter.ipynb:1]: The notebook generally looks good. Let's schedule a call to go over it.

File: vllm_hook_plugins/vllm_hook_plugins/analyzers/highlighter_analyzer.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
+import os
+import glob
+from typing import Any, Callable, Dict, Optional
+
+import torch
+from transformers import AutoModelForCausalLM
```
Comment (IRENEKO) [inline vllm_hook_plugins/vllm_hook_plugins/analyzers/highlighter_analyzer.py]: What is the role of the huggingface model here? Let's go over this as well in the discussion.

File: vllm_hook_plugins/vllm_hook_plugins/utils/TokenHighlighter/scorer_validation.py
Comment (IRENEKO) [inline vllm_hook_plugins/vllm_hook_plugins/utils/TokenHighlighter/scorer_validation.py:1]: Should this be kept locally as well for your own reference if this is a debugging script?

File: vllm_hook_plugins/vllm_hook_plugins/workers/highlighter_worker.py [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+
+from __future__ import annotations
+
+import os
+import warnings
+from typing import TYPE_CHECKING, Any
+
+import torch
+import torch.nn.functional as F
+from transformers import AutoModelForCausalLM, AutoTokenizer
```
Comment (IRENEKO) [inline vllm_hook_plugins/vllm_hook_plugins/workers/highlighter_worker.py]: Do we still need to load huggingface model in the worker?

File: vllm_hook_plugins/vllm_hook_plugins/_hook_plugin.py [hunk 8]
```diff
@@ -135,9 +135,12 @@ async def _patched_generate(
     wants_qk = extra.get("output_qk") is not None
     wants_steer = isinstance(extra.get("steer"), dict)
     needs_hooks = wants_hs or wants_qk or wants_steer
+    # Token Highlighter writes its own artifacts inside the wrapped execute_model;
+    # it needs install_hooks but not the probe-style flush/get below.
+    needs_highlighter = bool(extra.get("highlighter_mode"))
```
Comment (IRENEKO) [inline vllm_hook_plugins/vllm_hook_plugins/_hook_plugin.py]: I am reluctant to open a flag for any specific use case... Is there a specific reason why you can't reuse `needs_hooks`?

**Review by IRENEKO — COMMENTED** [review]
Thanks Arav for your PR revision! Please see the comments in the individual file for details.

File: examples/demo_token_highlighter.py (hunk 1 above)
Comment (asanth7) [reply examples/demo_token_highlighter.py]: This is a wrapper for highlighter-specific config/loading to avoid modifying `hook_llm.py`, but can be removed if we are able to edit that directly instead.

File: examples/demo_token_highlighter.py (hunk 4 above)
Comment (asanth7) [reply examples/demo_token_highlighter.py]: Sorry, this is an old artifact (should be `token_highlighter` after I'd refactored elsewhere). Thanks for pointing this out -- will fix it.

File: notebooks/demo_token_highlighter/demo_token_highlighter.ipynb
Comment (asanth7) [reply notebooks/demo_token_highlighter/demo_token_highlighter.ipynb:1]: Sounds good!

File: examples/demo_token_highlighter.py (hunk 5 above)
Comment (asanth7) [reply examples/demo_token_highlighter.py:45]: Prefix caching is possible -- we disable it here to ensure that "mitigate" mode uses the patched KV cache (it requires a fresh prefill with the beta-scaled embeddings separate from "capture" model prefill) although it is not strictly necessary. We reset prefix cache in the demo between prompts anyway. It seems more robust to disable prefix caching to ensure the capture/mitigate pipeline is most reliable, especially with repeated or identical prompts where prefixes are similar.

File: vllm_hook_plugins/vllm_hook_plugins/_hook_plugin.py (hunk 8 above)
Comment (asanth7) [reply vllm_hook_plugins/vllm_hook_plugins/_hook_plugin.py]: I included the `highlighter_mode` check within `need_hooks` so that we can re-use `need_hooks` (just added a `wants_highlighter` flag like with other workers) and followed a very similar structure as the steering worker; Token Highlighter handles its own hook artifacts (highlighter_activations.pt from the worker for score computation and highlighter.pt from the analyzer for mitigation) and so it doesn't require the post-processing/flushing that the probe method uses.

File: .gitignore
Comment (IRENEKO) [inline .gitignore:1]: Let's remove this file from the PR

File: TokenHighlighter.pdf
Comment (IRENEKO) [inline TokenHighlighter.pdf:1]: Please move this file to the respective folder following the PR template.

File: docs/use_cases/live_highlighter/TokenHighlighter_GradientScore_Derivation.pdf
Comment (IRENEKO) [inline docs/use_cases/live_highlighter/TokenHighlighter_GradientScore_Derivation.pdf:1]: Please either consolidate this doc with the TokenHighlighter.pdf or remove from the PR as we have discussed.

File: vllm_hook_plugins/vllm_hook_plugins/hook_llm.py
Comment (IRENEKO) [inline vllm_hook_plugins/vllm_hook_plugins/hook_llm.py:1]: Not sure if we have touched on this topic, but is there a way to enable token highlighter without modifying `hook_llm.py` and `_hook_plugin.py`? I think the other use case `https://github.com/IBM/vLLM-Hook/blob/main/docs/use_cases/spotlight.md` faces similar challenges and was able to avoid modifying the core files. Can you please take a look?

**Review by IRENEKO — CHANGES_REQUESTED** [review]

File: vllm_hook_plugins/vllm_hook_plugins/hook_llm.py
Comment (IRENEKO) [inline vllm_hook_plugins/vllm_hook_plugins/hook_llm.py:1]: Is this revision needed?

File: notebooks/demo_token_highlighter/demo_token_highlighter.ipynb
Comment (IRENEKO) [inline notebooks/demo_token_highlighter/demo_token_highlighter.ipynb:1]: Quick verification: have you re-run the notebook after your latest commit?

File: notebooks/demo_token_highlighter/demo_token_highlighter_colab.ipynb
Comment (IRENEKO) [inline notebooks/demo_token_highlighter/demo_token_highlighter_colab.ipynb:1]: Just to confirm if you have run and verified the colab notebook as I didn't see cell outputs

File: docs/use_cases/live_highlighter/TokenHighlighter_GradientScore_Derivation.pdf
Comment (IRENEKO) [inline docs/use_cases/live_highlighter/TokenHighlighter_GradientScore_Derivation.pdf:1]: This comment is not resolved yet.

File: docs/use_cases/live_highlighter/TokenHighlighter.pdf
Comment (IRENEKO) [inline docs/use_cases/live_highlighter/TokenHighlighter.pdf:1]: This seems to be a duplicate of the markdown file

**Review by IRENEKO — CHANGES_REQUESTED** [review]

File: notebooks/demo_token_highlighter/demo_token_highlighter.ipynb
Comment (asanth7) [reply notebooks/demo_token_highlighter/demo_token_highlighter.ipynb:1]: Yes, the only major difference was the refactoring to use the new generate and analyze wrappers (output and internals behave the same).

**Review by IRENEKO — APPROVED** [review]
